firefox_puppeteer/ui/windows.py

```diff
@@ -0,0 +1,479 @@
+        handles_to_keep = [entry.handle if isinstance(entry, BaseWindow) else entry
+                           for entry in exceptions]
+
+        for handle in self.marionette.chrome_window_handles:
+            if handle not in handles_to_keep:
+                self.close(handle)
+
+    def create_window_instance(self, handle, expected_class=None):
+        """Creates a :class:`BaseWindow` instance for the given chrome window
+
+        :param handle: The handle of the chrome window
+        :param expected_class: Optional, check for the correct window class
+        """
+        current_handle = self.marionette.current_chrome_window_handle
+
+        try:
+            # Retrieve window type to determine the type of window
+            self.switch_to(handle)
+            window_type = self.marionette.get_window_type()
+        finally:
+            # Ensure to switch back to the original window
+            if handle != current_handle:
+                self.marionette.switch_to_window(current_handle)
+
+        window = BaseWindow(lambda: self.marionette, handle)
+
+        if window_type == 'navigator:browser':
+            window.__class__ = BrowserWindow
+        else:
+            raise errors.UnknownWindowError('Unknown window type "%s" for handle: "%s"' %
+                                            (window_type, handle))
+
+        if expected_class is not None and type(window) is not expected_class:
+            raise errors.UnexpectedWindowTypeError('Expected window "%s" but got "%s"' %
+                                                   (expected_class, type(window)))
+
+        return window
+
+    def focus(self, handle):
+        """Focuses the chrome window with the given handle.
+
+        :param handle: The handle of the chrome window
+        """
+        self.switch_to(handle)
+
+        with self.marionette.using_context('chrome'):
+            self.marionette.execute_script(""" window.focus(); """)
+
+        wait = Wait(self.marionette, timeout=5, interval=.1)
+        wait.until(lambda m: handle == self.focused_chrome_window_handle)
+
+    def switch_to(self, target):
+        """Switches context to the specified chrome window.
+
+        :param target: The window to switch to. `target` can be a `handle` or a
+                       callback that returns True in the context of the desired
+                       window.
+
+        :returns: The old `window_handle`. This makes it easy to switch back
+                  to the original window later.
+        """
+        target_handle = None
+
+        if target in self.marionette.chrome_window_handles:
+            target_handle = target
+        elif callable(target):
+            current_handle = self.marionette.current_chrome_window_handle
+
+            # switches context if callback for a chrome window returns `True`.
+            for handle in self.marionette.chrome_window_handles:
+                self.marionette.switch_to_window(handle)
```

Comment:
You have to when you want to run the callback against the actual chrome window. Otherwise it would not work on the target window but the currently selected one, and any kind of element retrieval would fail.